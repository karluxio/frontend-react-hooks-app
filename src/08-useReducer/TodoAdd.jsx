import { useForm } from "../hooks/useForm"

export const TodoAdd = ({ onNewTodo }) => {

  const { description, onInputChange, onResetForm } = useForm({ description: '' })

  const onSubmit = (e) => {
    e.preventDefault()

    if (description <= 1) return
    onNewTodo({
      id: new Date().getTime(),
      description,
      done: false
    })

    onResetForm()
  }

  return (
    <>
      <h4>Add todo</h4>
      <hr />
      <form onSubmit={onSubmit}>
        <input
          className="form-control"
          name="description"
          onChange={onInputChange}
          placeholder="cleanup my house"
          type="text"
          value={description}
        />

        <button type="submit" className="btn btn-outline-primary mt-2">Add</button>
      </form>
    </>
  )
}