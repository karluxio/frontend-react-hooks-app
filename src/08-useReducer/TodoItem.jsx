export const TodoItem = ({ todo, onDeleteTodo, onToggleTodo }) => {
  return (
    <li className="d-flex justify-content-between list-group-item">

      <span
        onClick={() => onToggleTodo(todo.id)}
        className={`align-self-center ${(todo.done) ? 'text-decoration-line-through' : 'text-decoration-none'}`}
      >
        {todo.description}
      </span>

      <button
        onClick={() => onDeleteTodo(todo.id)}
        className="btn btn-danger">
        delete
      </button>

    </li >
  )
}