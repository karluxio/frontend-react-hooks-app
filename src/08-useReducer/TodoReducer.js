
export const initialState = []

export const todoReducer = (state = initialState, { type, payload }) => {
  switch (type) {

    case '[TODO] add':
      return [...state, { ...payload }]

    case '[TODO] toggle':
      return [...state].map(todo => {
        if (todo.id === payload) {
          return {
            ...todo,
            done: !todo.done
          }
        }
        return todo
      })

    case '[TODO] delete':
      return [...state].filter(todo => todo.id !== payload)

    default:
      return state
  }
}
