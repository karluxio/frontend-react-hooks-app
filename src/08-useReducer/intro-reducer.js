export const initialState = [
  {
    id: 1,
    todo: 'recollect the soul gem',
    done: false
  }
]

export const todosReducer = (state = initialState, action = {}) => {

  if (action.type === "[TODO] add todo") {
    return [...state, action.payload]
  }

  return state
}

let todos = todosReducer()

const newTodo = {
  id: 2,
  todo: "recollect the power gem",
  done: true
}

const addTodoAction = {
  type: '[TODO] add todo',
  payload: newTodo
}

todos = todosReducer(todos, addTodoAction)

console.log({ todos });
