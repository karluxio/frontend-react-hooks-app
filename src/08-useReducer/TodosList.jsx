import { TodoItem } from "./TodoItem"

export const TodosList = ({ todos = [], onDeleteTodo, onToggleTodo }) => {
  return (
    <ul className="list-group">
      {
        todos.length > 0 && todos.map(todo => (
          <TodoItem key={todo.id} todo={todo} onDeleteTodo={onDeleteTodo} onToggleTodo={onToggleTodo} />
        ))
      }
    </ul>
  )
}