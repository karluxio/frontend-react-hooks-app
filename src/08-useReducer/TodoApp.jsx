import { useReducer, useEffect } from "react"
import { todosReducer, initialState } from "./TodoReducer"
import { TodosList } from "./TodosList"
import { TodoAdd } from "./TodoAdd"

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || []
}

export const TodoApp = () => {

  const [todos, dispatch] = useReducer(todosReducer, initialState, init)

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos))
  }, [todos])


  const handleNewTodo = (todo) => {
    dispatch({ type: '[TODO] add', payload: todo })
  }

  const handleDeleteTodo = (id) => {
    dispatch({ type: '[TODO] delete', payload: id })
  }

  const handleToggleTodo = (id) => {
    dispatch({ type: '[TODO] toggle', payload: id })
  }

  return (
    <>

      <h1>TodoApp: {todos.length}, <small>pending: 4</small></h1>
      <hr />

      <div className="row">

        <div className="col-7">
          <TodosList todos={todos} onDeleteTodo={handleDeleteTodo} onToggleTodo={handleToggleTodo} />
        </div>

        <div className="col-5">
          <TodoAdd onNewTodo={handleNewTodo} />
        </div>

      </div>

    </>
  )
}