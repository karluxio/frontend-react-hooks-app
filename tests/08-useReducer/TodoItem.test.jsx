import { render, screen, fireEvent } from '@testing-library/react'
import { TodoItem } from '../../src/08-useReducer/TodoItem'

describe('TodoItem', () => {

  const onDeleteTodoMock = jest.fn()
  const onToggleTodoMock = jest.fn()
  const todo = {
    id: 1,
    description: 'some description',
    done: false
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("should match against snapshot", () => {
    const { container } = render(
      <TodoItem
        todo={todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it("should call onToggleTodo when click on toggle button", async () => {
    render(
      <TodoItem
        todo={todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    )

    const span = screen.getByText(todo.description)
    fireEvent.click(span)
    expect(onToggleTodoMock).toHaveBeenCalledTimes(1)
    expect(onToggleTodoMock).toHaveBeenCalledWith(todo.id)
  })

  it("should call onDeleteTodo when click on delete button", async () => {
    render(
      <TodoItem
        todo={todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    )

    const button = screen.getByRole('button', { name: /delete/i })
    fireEvent.click(button)
    expect(onDeleteTodoMock).toHaveBeenCalledTimes(1)
    expect(onDeleteTodoMock).toHaveBeenCalledWith(todo.id)
  })

  it("should has a className text-decoration-none", async () => {
    render(
      <TodoItem
        todo={todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    )

    const span = screen.getByText(todo.description)
    expect(span.className).toContain('text-decoration-none')
    expect(span.className).not.toContain('text-decoration-line-through')
  })

  it("should has a className text-decoration-line-through", async () => {

    todo.done = true

    render(
      <TodoItem
        todo={todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    )

    const span = screen.getByText(todo.description)
    expect(span.className).toContain('text-decoration-line-through')
    expect(span.className).not.toContain('text-decoration-none')
  })
})