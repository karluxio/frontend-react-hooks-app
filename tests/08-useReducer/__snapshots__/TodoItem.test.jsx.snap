// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should match against snapshot 1`] = `
<div>
  <li
    class="d-flex justify-content-between list-group-item"
  >
    <span
      class="align-self-center text-decoration-none"
    >
      some description
    </span>
    <button
      class="btn btn-danger"
    >
      delete
    </button>
  </li>
</div>
`;
